### PR TITLE
busybox: update to 1.37.0

### PR DESCRIPTION
**Pull request by rsalvaterra**
Add a patch to fix non-x86 builds, refresh the configuration and patches.

Signed-off-by: Rui Salvaterra \<rsalvaterra@gmail.com\>

### CODE REVIEW

**Comment by rsalvaterra** [comment]
SHA hardware acceleration detection seems to be bugged, causing the build to fail on non-x86_64, will take a look at it later in the day.

**Comment by hnyman** [comment]
It looks a bit untypical that there are no more wording changes to the Config.in files.

Would be nice if you verify that there are no other changes in the various Config.in files.
The typical actions needed at busybox major version upgrades are documented  in my previous PRs (e.g. to 1.36.0 and 1.35.0)
#11684 or #4891 
(the automatic config file update logic destroys some of our modifications, so they need to be re-applied after running the update scripts)

**Comment by rsalvaterra** [comment]
Thanks for the hints, @hnyman! Actually, I didn't run any scripts, I did everything manually, except for the patch refreshing, of course.

**Comment by rsalvaterra** [comment]
Ok, I think this is ready to drop the draft status. @hnyman, mind taking a look, if/when you have time?

File: package/utils/busybox/Config-defaults.in [hunk 1]
```diff
@@ -2239,8 +2248,7 @@ config BUSYBOX_DEFAULT_FEATURE_WATCHDOG_OPEN_TWICE
 	default n
 config BUSYBOX_DEFAULT_FEATURE_IPV6
 	bool
-	default y if IPV6
-	default n
+	default y
```
Comment (hnyman) [inline package/utils/busybox/Config-defaults.in:2251]: This looks wrong. You have not yet returned the OpenWrt logic here. One of the corrections mentioned in my PR comments.

**Comment by hnyman** [comment]
At the first glance, I spotted just the one mistake that I commented.
Now the changes look more like I expected, lots of minor changes in the config,in files.

You might also document the update script and corrections done to comment message. Helps the next time. (You get some from shell history, and other parts you can largely copy from my old pr.)

I can look closer on Sunday when I get home.

File: package/utils/busybox/Config-defaults.in (hunk 1 above)
Comment (rsalvaterra) [reply package/utils/busybox/Config-defaults.in:2251]: Actually, I thought `CONFIG_IPV6` had been completely dropped, but it seems the symbol still exists. Will fix, thanks.

**Comment by rsalvaterra** [comment]
Just a heads up, for some reason, this seems to be interacting badly with hostapd, none of my interfaces are coming up. I have no time to debug this at the moment, but possibly during the weekend.

File: package/utils/busybox/Config-defaults.in [hunk 2]
```diff
@@ -2240,7 +2249,6 @@ config BUSYBOX_DEFAULT_FEATURE_WATCHDOG_OPEN_TWICE
 config BUSYBOX_DEFAULT_FEATURE_IPV6
 	bool
 	default y if IPV6
-	default n
```
Comment (hnyman) [inline package/utils/busybox/Config-defaults.in:2243]: You still need to leave the fallback default "no" there... (You are now removing it)

Comment (rsalvaterra) [reply package/utils/busybox/Config-defaults.in]: Yeah, stupid oversight on my part. Fixed now, thanks!

**Comment by hnyman** [comment]
There is still one unexpected change of a recent change to default.
You are reverting 86e3016896f probably by accident.

**Comment by rsalvaterra** [comment]
> There is still one unexpected change of a recent change to default. You are reverting [86e3016](https://github.com/openwrt/openwrt/commit/86e3016896f43fdcd1ee34a8aafe92b2dab9367e) probably by accident.

Thanks, also fixed now.

**Comment by rsalvaterra** [comment]
Commit message changed to include the full procedure.

**Comment by hnyman** [comment]
I made a test build for my mt7622/RT3200 and noticed two problems.

(I first compiled  r27611-4eaa254018 without changes, and then the same revision with the new busybox applied.)

1) like you mention earlier, there is something wrong with hostapd. The 5 GHz radio did not come properly up.

```
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: Set new config for phy wl1: /var/run/hostapd-wl1.conf
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: Restart interface for phy wl1
Mon Sep 30 20:39:50 2024 daemon.err odhcp6c[4025]: Failed to send SOLICIT message to ff02::1:2 (Address not available)
Mon Sep 30 20:39:50 2024 user.notice ucitrack: Setting up /etc/config/luci-splash reload dependency on /etc/config/firewall
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: Configuration file: data: driver=nl80211 logger_syslog=127 log....
Mon Sep 30 20:39:50 2024 user.notice ucitrack: Setting up /etc/config/qos reload dependency on /etc/config/firewall
Mon Sep 30 20:39:50 2024 user.notice ucitrack: Setting up /etc/config/miniupnpd reload dependency on /etc/config/firewall
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.500277] br-lan: port 6(hn5wpa3) entered blocking state
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.505782] br-lan: port 6(hn5wpa3) entered disabled state
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.511390] mt7915e 0000:01:00.0 hn5wpa3: entered allmulticast mode
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.517952] mt7915e 0000:01:00.0 hn5wpa3: entered promiscuous mode
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.524312] br-lan: port 6(hn5wpa3) entered blocking state
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.529856] br-lan: port 6(hn5wpa3) entered forwarding state
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hn5wpa3: interface state UNINITIALIZED->COUNTRY_UPDATE
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.539081] br-lan: port 6(hn5wpa3) entered disabled state
Mon Sep 30 20:39:50 2024 daemon.err hostapd: Driver does not support configured HT capability [RX-STBC*]
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hn5wpa3: interface state COUNTRY_UPDATE->DISABLED
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hn5wpa3: AP-DISABLED
Mon Sep 30 20:39:50 2024 daemon.err hostapd: hn5wpa3: Unable to setup interface.
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: nl80211: deinit ifname=hn5wpa3 disabled_11b_rates=0
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.558720] mt7915e 0000:01:00.0 hn5wpa3: left allmulticast mode
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.564804] mt7915e 0000:01:00.0 hn5wpa3: left promiscuous mode
Mon Sep 30 20:39:50 2024 kern.info kernel: [   24.570952] br-lan: port 6(hn5wpa3) entered disabled state
Mon Sep 30 20:39:50 2024 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hn5wpa3: CTRL-EVENT-TERMINATING
Mon Sep 30 20:39:50 2024 daemon.err hostapd: hostapd_free_hapd_data: Interface hn5wpa3 wasn't started
Mon Sep 30 20:39:50 2024 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hn5wpa2r: CTRL-EVENT-TERMINATING
Mon Sep 30 20:39:50 2024 daemon.err hostapd: hostapd_free_hapd_data: Interface hn5wpa2r wasn't started
Mon Sep 30 20:39:50 2024 daemon.notice hostapd: hostapd.add_iface failed for phy wl1 ifname=hn5wpa3
```

2. The second thing is cron: It does not execute the root crontab file properly. Instead, it throws a log item each minute:
```

root@router4:/etc/luci_statistics# cat /etc/crontabs/root
0 2 * * * /etc/init.d/luci_statistics backup
```

```
root@router4:/etc/luci_statistics# logread | grep cron
Mon Sep 30 20:39:45 2024 cron.err crond[2633]: crond (busybox 1.37.0) started, log level 5
Mon Sep 30 20:39:45 2024 cron.err crond[2633]: user:root entry:(null)
Mon Sep 30 20:40:44 2024 cron.err crond[2633]: user:root entry:(null)
Mon Sep 30 20:40:44 2024 cron.err crond[2633]: wakeup dt=59
Mon Sep 30 20:40:44 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:40:44 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
Mon Sep 30 20:41:00 2024 cron.err crond[2633]: wakeup dt=16
Mon Sep 30 20:41:00 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:41:00 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
Mon Sep 30 20:42:00 2024 cron.err crond[2633]: wakeup dt=60
Mon Sep 30 20:42:00 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:42:00 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
Mon Sep 30 20:43:00 2024 cron.err crond[2633]: wakeup dt=60
Mon Sep 30 20:43:00 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:43:00 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
Mon Sep 30 20:44:00 2024 cron.err crond[2633]: wakeup dt=60
Mon Sep 30 20:44:00 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:44:00 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
Mon Sep 30 20:45:00 2024 cron.err crond[2633]: wakeup dt=60
Mon Sep 30 20:45:00 2024 cron.err crond[2633]: file root:
Mon Sep 30 20:45:00 2024 cron.err crond[2633]:  line /etc/init.d/luci_statistics backup
```

Reverting back to the build woth bb 1.36.0 makes the hostapd to work ok and cron to stop complaining. So clearly something related to bb 1.37.0


Ps. As a sidenote, I made one further test: sysupgrade from the new busybox is possible...  (unlike is was with 1.34.0 where an upstream bug caused trouble for us https://github.com/openwrt/openwrt/pull/4488#issuecomment-908459881 )

**Comment by rsalvaterra** [comment]
Yeah, at this point it's probably better just leaving this as draft and keeping an eye out in the BusyBox mailing list, for the time being. It's a bit perplexing that a tool that is clearly targeting embedded systems gets a major release with so many regressions (and not even building on anything other than x86). Thoughts?

**Comment by hnyman** [comment]
The crond error might actually just be an annoying logging change, as the only recent change to bb cron code is about the error level in one logging function.

The hostapd thing, disabled wifi interface(s) might be more difficult to track.

**Comment by hnyman** [comment]
The crond log is just about logging change:

I tested it by taking the only crond change in 1.37.0 ( https://git.busybox.net/busybox/commit/miscutils/crond.c?id=01f2b5976d03fb50229d28297cf5ebd8be533203 ) as patch to 1.36.1, and the same log items then started to appear there. The patch changes the earlier wrong logging level from 4 to 5 in "level 5 logging" function.

The change is brought to daylight by our init script /etc/init.d/crond setting cron logging threshold to 5, which now causes the earlier "level 4" info message to get printed as they are now level 5.  The init script has been that way at least since year 2013, but due to the upstream bug, the level 5 has been meaningless earlier (and only level 7 and 8 messages have been printed).

To get back to the old logging contents, we should probably change the crond init script to specify 6 as default log_level (so that it is once again one step  above the level 4/5).

**Comment by hnyman** [comment]
@rsalvaterra 
Have you got any ideas about the culprit for the hostapd related anomalies?
Thinking how busybox may have impact on hostapd related stuff, the main thing that pops into mind is shell execution of the init script / hotplug script / something. Then looking at the ash changes between 1.36 and 1.37, there aren't that many obvious suspect looking commits:  https://git.busybox.net/busybox/log/shell/ash.c

By just a hunch, without testing anything, my favourite suspect is "ash: fix handling of single-quoted strings in pattern substitution"  https://git.busybox.net/busybox/commit/shell/ash.c?id=c5a1be25ba6dd705382ce6c25d96a32f79974c04

Ps.
You might change the crond init script to specify log level 6 like I suggested.

**Comment by Neustradamus** [comment]
@rsalvaterra: Have you seen @hnyman comments since your latest comment?

**Comment by hauke** [comment]
It is broken since this commit: https://github.com/mirror/busybox/commit/c1c267fd36b0fcac8c8871232eecc1e360173990

**Comment by hnyman** [comment]
> broken since this commi

Interesting.
Did you figure out, where in our code the changes affect us? In hostapd init script?
I.e. is there an upstream bug, or did the changes just bring some our problem into highlight?

**Comment by rsalvaterra** [comment]
Sorry, everyone, I haven't been able to pay much attention to this (or any OpenWrt-related work) in the last weeks. I'm moving flats, so all hobbies had to be put on the back burner. :smiling_face_with_tear:

**Comment by hnyman** [comment]
> broken since this commit: https://github.com/mirror/busybox/commit/c1c267fd36b0fcac8c8871232eecc1e360173990

Looking at the upstream busybox commit where @hauke thinks that things break, most of the changes in math.c concerns conditional code inside `#if ENABLE_FEATURE_SH_MATH_BASE`, which is not true by default in our busybox. That would be all lines until ~600 when the if segment ends.

Only the last changes to lines 700+ do get included in our code. And there is a change in handling out-of-range numbers.

```
@@ -702,7 +721,6 @@ evaluate_string(arith_state_t *math_state, const char *expr)
 				dbg("[%d] var:IGNORED", (int)(numstackptr - numstack));
 				expr = p;
 				numstackptr->var_name = NULL;
- push_num0:
 				numstackptr->val = 0;
 			}
  push_num:
@@ -715,11 +733,12 @@ evaluate_string(arith_state_t *math_state, const char *expr)
 			/* Number */
 			char *end;
 			numstackptr->var_name = NULL;
-			errno = 0;
 			/* code is smaller compared to using &expr here: */
 			numstackptr->val = strto_arith_t(expr, &end);
 			expr = end;
 			dbg("[%d] val:%lld", (int)(numstackptr - numstack), numstackptr->val);
+			if (!expr) /* example: $((10#)) */
+				goto syntax_err;
 			/* A number can't be followed by another number, or a variable name.
 			 * We'd catch this later anyway, but this would require numstack[]
 			 * to be ~twice as deep to handle strings where _every_ char is
@@ -728,13 +747,6 @@ evaluate_string(arith_state_t *math_state, const char *expr)
 			 */
 			if (isalnum(*expr) || *expr == '_')
 				goto syntax_err;
-			if (errno) {
-// TODO: bash 5.2.15 does not catch ERANGE (some older version did?).
-// $((99999999999999999999)) is 7766279631452241919 (the 64-bit truncated value).
-// Our BASE# code does this as well: try $((10#99999999999999999999)),
-// but the "ordinary" code path with strtoull() does not do this.
-				goto push_num0; /* bash compat */
-			}
 			goto push_num;
 		}
 
```
And meanwhile the upstream code has evolved a bit and comments have been added:
```
				dbg("[%d] var:IGNORED", (int)(numstackptr - numstack));
				expr = p;
				numstackptr->var_name = NULL; /* not needed, paranoia */
				numstackptr->val = 0; /* not needed, paranoia */
			}
 push_value:
			numstackptr++;
			lasttok = TOK_VALUE;
			continue;
		}

		if (isdigit(*expr)) {
			/* Number */
			char *end;
			numstackptr->var_name = NULL;
			/* code is smaller compared to using &expr here: */
			numstackptr->val = strto_arith_t(expr, &end);
			expr = end;
			dbg("[%d] val:%lld", (int)(numstackptr - numstack), numstackptr->val);
			if (!expr) /* example: $((10#)) */
				goto syntax_err;
			/* A number can't be followed by another number, or a variable name.
			 * We'd catch this later anyway, but this would require numstack[]
			 * to be ~twice as deep to handle strings where _every_ char is
			 * a new number or name.
			 * Examples: "09" is two numbers, "0v" is number and name.
			 */
			if (isalnum(*expr) || *expr == '_')
				goto syntax_err;
			goto push_value;
		}
```
I wonder if we should try reviving that errno handling.

**Comment by hauke** [comment]
I am testing this with mac80211-hwsim in qemu. 

I am using this configuration:
```
uci set wireless.radio0.channel=11
uci set wireless.radio0.band=2g
uci delete wireless.radio0.disabled
uci set wireless.radio0.ldpc=0
uci set wireless.radio0.rx_stbc=0
uci set wireless.radio0.max_amsdu=0
uci set wireless.default_radio0.encryption=sae-mixed
uci set wireless.default_radio0.key=testtest
uci delete wireless.radio1.channel
uci set wireless.radio1.band=2g
uci delete wireless.radio1.disabled
uci set wireless.default_radio1.network=wan
uci set wireless.default_radio1.mode=sta
uci set wireless.default_radio1.encryption=sae-mixed
uci set wireless.default_radio1.key=testtest

uci commit
service network reload
```

It is failing with:
```
Mon Nov 18 22:02:36 2024 daemon.err hostapd: Driver does not support configured HT capability [TX-STBC]
```

This is the diff of the `/var/run/hostapd-phy0.conf` file:
```
@@ -368,7 +383,7 @@
 #num_global_macaddr=1
 ieee80211n=1
 ht_coex=0
-ht_capab=[HT40-][SHORT-GI-20][SHORT-GI-40][TX-STBC][DSSS_CCK-40]
+ht_capab=[HT40-][SHORT-GI-20][SHORT-GI-40][DSSS_CCK-40]
 ieee80211ax=1
 he_bss_color=128
 he_spr_sr_control=3
```
See this file: https://github.com/openwrt/openwrt/blob/main/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh#L224

**Comment by hnyman** [comment]
My error log above had slightly different error, other capability, but same place...

`daemon.err hostapd: Driver does not support configured HT capability [RX-STBC*]`

I guess that somehow the capability bit field overflows, or something like that?

**Comment by hnyman** [comment]
Pure guess, but it could be something related to these bitshifts:

https://github.com/openwrt/openwrt/blob/b0312c10818f36bc6e61a518bb528d2809562796/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh#L215-L217

https://github.com/openwrt/openwrt/blob/b0312c10818f36bc6e61a518bb528d2809562796/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh#L350-L352

**Comment by hnyman** [comment]
Thanks @hauke 
I compiled with your patch, and it works.

@rsalvaterra 
Are you going to update this PR?  With the cron init file change and the patch from Hauke?

**Comment by hauke** [comment]
This PR fixes the problem I ran into with the busybox upgrade: https://github.com/openwrt/openwrt/pull/17043

**Comment by hauke** [comment]
This script reproduces the problem in busybox:
```
#!/bin/sh

phy=phy0

set -x 

ht_cap_mask=0
for cap in $(iw phy "$phy" info | grep 'Capabilities:' |grep -v PHY  | cut -d: -f2); do
	echo $cap
	ht_cap_mask="$(($ht_cap_mask | $cap))"
done

echo $ht_cap_mask
```

I think we should only fix the script and do not patch busybox. I am a bit concerned that we may have more such problems.

**Comment by hnyman** [comment]
Nice debugging.
But you are rightly concerned. Apparently the busybox change just exposed a bug in our own code, like we feared.

**Comment by hnyman** [comment]
@rsalvaterra 
Could you please update your PR with the needed cron init script change?
(level 5 changed to 6)

```
--- a/package/utils/busybox/files/cron
+++ b/package/utils/busybox/files/cron
@@ -28,7 +28,7 @@ start_service() {
 	ln -s /etc/crontabs /var/spool/cron/ 2>/dev/null
 
 	procd_open_instance
-	procd_set_param command "$PROG" -f -c /etc/crontabs -l "${loglevel:-5}"
+	procd_set_param command "$PROG" -f -c /etc/crontabs -l "${loglevel:-6}"
 	for crontab in /etc/crontabs/*; do
 		 procd_set_param file "$crontab"
 	done

```

With that change and the patch from @hauke in PR #17043 , the busybox 1.37.0 might be ready for wider testing in main/master.

**Comment by hauke** [comment]
I used this script to test the problem:
```
#!/bin/sh

phy=phy0

set -x 

ht_cap_mask=0
for cap in $(iw phy "$phy" info | grep 'Capabilities:' | cut -d: -f2); do
	echo $cap
	ht_cap_mask="$(($ht_cap_mask | $cap))"
done

echo $ht_cap_mask
```

This is the iw output:
```
root@OpenWrt:~# iw phy phy0 info | grep 'Capabilities:' 
		Capabilities: 0x107e
			HE PHY Capabilities: (0x02bfce0000000000000000):
			EHT PHY Capabilities: (0x7c0000feffff7f01):
			HE PHY Capabilities: (0x02bfce0000000000000000):
			EHT PHY Capabilities: (0x7c0000feffff7f01):
			HE PHY Capabilities: (0x02bf000000000000000000):
		Capabilities: 0x107e
			HE PHY Capabilities: (0x1cbfce0000000000000000):
			EHT PHY Capabilities: (0xfc1f3ffeffff7f37):
			HE PHY Capabilities: (0x1cbfce0000000000000000):
			EHT PHY Capabilities: (0xfc1f3ffeffff7f37):
			HE PHY Capabilities: (0x1cbf000000000000000000):
			HE PHY Capabilities: (0x1cbfce0000000000000000):
			EHT PHY Capabilities: (0xfefffffeffffff7f):
			HE PHY Capabilities: (0x1cbfce0000000000000000):
			EHT PHY Capabilities: (0xfefffffeffffff7f):
			HE PHY Capabilities: (0x1cbf000000000000000000):
		Capabilities: 0x107e
```

With busybox 1.36.1 it returned this outout:
```
root@OpenWrt:~# /tmp/test.sh
+ ht_cap_mask=0
+ iw phy phy0 info
+ grep Capabilities:
+ cut -d: -f2
+ echo 0x107e
0x107e
+ ht_cap_mask=4222
+ echo '(0x02bfce0000000000000000)'
(0x02bfce0000000000000000)
+ ht_cap_mask=4222
+ echo '(0x7c0000feffff7f01)'
(0x7c0000feffff7f01)
+ ht_cap_mask=8935142755919691647
+ echo '(0x02bfce0000000000000000)'
(0x02bfce0000000000000000)
+ ht_cap_mask=8935142755919691647
+ echo '(0x7c0000feffff7f01)'
(0x7c0000feffff7f01)
+ ht_cap_mask=8935142755919691647
+ echo '(0x02bf000000000000000000)'
(0x02bf000000000000000000)
+ ht_cap_mask=8935142755919691647
+ echo 0x107e
0x107e
+ ht_cap_mask=8935142755919691647
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=8935142755919691647
+ echo '(0xfc1f3ffeffff7f37)'
(0xfc1f3ffeffff7f37)
+ ht_cap_mask=-279434287424503937
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-279434287424503937
+ echo '(0xfc1f3ffeffff7f37)'
(0xfc1f3ffeffff7f37)
+ ht_cap_mask=-279434287424503937
+ echo '(0x1cbf000000000000000000)'
(0x1cbf000000000000000000)
+ ht_cap_mask=-279434287424503937
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-279434287424503937
+ echo '(0xfefffffeffffff7f)'
(0xfefffffeffffff7f)
+ ht_cap_mask=-72057598332895361
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-72057598332895361
+ echo '(0xfefffffeffffff7f)'
(0xfefffffeffffff7f)
+ ht_cap_mask=-72057598332895361
+ echo '(0x1cbf000000000000000000)'
(0x1cbf000000000000000000)
+ ht_cap_mask=-72057598332895361
+ echo 0x107e
0x107e
+ ht_cap_mask=-72057598332895361
+ echo -72057598332895361
-72057598332895361
root@OpenWrt:~#
```

With busybox 1.37.0 it returned this:
```
root@OpenWrt:~# /tmp/test.sh
+ ht_cap_mask=0
+ iw phy phy0 info
+ grep Capabilities:
+ cut -d: -f2
+ echo 0x107e
0x107e
+ ht_cap_mask=4222
+ echo '(0x02bfce0000000000000000)'
(0x02bfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0x7c0000feffff7f01)'
(0x7c0000feffff7f01)
+ ht_cap_mask=-1
+ echo '(0x02bfce0000000000000000)'
(0x02bfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0x7c0000feffff7f01)'
(0x7c0000feffff7f01)
+ ht_cap_mask=-1
+ echo '(0x02bf000000000000000000)'
(0x02bf000000000000000000)
+ ht_cap_mask=-1
+ echo 0x107e
0x107e
+ ht_cap_mask=-1
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0xfc1f3ffeffff7f37)'
(0xfc1f3ffeffff7f37)
+ ht_cap_mask=-1
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0xfc1f3ffeffff7f37)'
(0xfc1f3ffeffff7f37)
+ ht_cap_mask=-1
+ echo '(0x1cbf000000000000000000)'
(0x1cbf000000000000000000)
+ ht_cap_mask=-1
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0xfefffffeffffff7f)'
(0xfefffffeffffff7f)
+ ht_cap_mask=-1
+ echo '(0x1cbfce0000000000000000)'
(0x1cbfce0000000000000000)
+ ht_cap_mask=-1
+ echo '(0xfefffffeffffff7f)'
(0xfefffffeffffff7f)
+ ht_cap_mask=-1
+ echo '(0x1cbf000000000000000000)'
(0x1cbf000000000000000000)
+ ht_cap_mask=-1
+ echo 0x107e
0x107e
+ ht_cap_mask=-1
+ echo -1
-1
root@OpenWrt:~# 
```

**Comment by hauke** [comment]
I merged the shell script fix in adf958c9192d94fbeeb0ec8e16800b0071455909, please rebase on top of it.

**Comment by hnyman** [comment]
Rebased version of this PR in #17107

**Comment by hauke** [comment]
Merged using #17107 in 89bb4050f7c555db4738f52d22f51ddb94875d4f and fc0c518a88e68d3deef04bec73b33d35186d6546.

**Pull request closed without merging**